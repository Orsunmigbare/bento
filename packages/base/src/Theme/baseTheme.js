import { ThemeManager } from "@staccx/theme"

export const theme = {
  name: "Default theme",
  spacing: {
    huge: "96px",
    large: "48px",
    mediumPlus: "32px",
    medium: "24px",
    small: "12px",
    tiny: "6px",
    micro: "3px"
  },
  wrapper: {
    small: "420px",
    medium: "640px",
    large: "1024px",
    breakout: "640px"
  },
  color: {
    white: "#fff",
    bg: "#fff",
    subtleHover: "#F7F7FF",
    bgGray: "#f7f7ff",
    line: "#EBEBF2",
    gray: "#767676",
    wcag: "#777",
    disabled: "#BABABA",
    primary: "#00a190",
    secondary: "#F5A623",
    text: "#2b2b2b",
    negative: "#ea526f",
    positive: "#4A90E2",
    warning: "#c54059",
    black: "#000",
    green: "#00a190" // for toggle eks
  },
  fontFamily: {
    heading: "sans-serif",
    body: "sans-serif"
  },
  font: {
    h1: "36px",
    h2: "26px",
    h3: "22px",
    h4: "18px",
    h5: "16px",
    h6: "14px",
    base: "14px",
    tiny: "12px",
    input: "16px",
    huge: "44px"
  },
  targetSize: {
    small: "36px",
    normal: "44px"
  },
  borderRadius: "2px"
}

const manager = new ThemeManager(theme)
export default manager
