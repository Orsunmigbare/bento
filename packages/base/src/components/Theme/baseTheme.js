export const theme = {
  name: "StaccXBase",
  spacing: {
    huge: "96px",
    large: "48px",
    mediumPlus: "32px",
    medium: "24px",
    small: "12px",
    tiny: "6px"
  },
  wrapper: {
    small: "420px",
    medium: "640px",
    large: "1024px"
  },
  color: {
    bg: "#fff",
    bgGray: "#f7f7ff",
    primary: "#00a190",
    gray: "#767676",
    text: "#2b2b2b",
    white: "#fff",
    red: "#ea526f",
    green: "#00a190",
    blue: "#988EFB",
    yellow: "#f4da87",
    line: "#EBEBF2",
    wcag: "#777",
    disabled: "#BABABA",
    warning: "#c54059",
    black: "#000",
    subtleHover: "#F7F7FF"
  },
  font: {
    size: {
      h1: "36px",
      h2: "26px",
      h3: "22px",
      h4: "18px",
      h5: "16px",
      h6: "14px",
      base: "14px",
      tiny: "12px",
      input: "16px",
      huge: "44px"
    },
    heading: "Libre Baskerville",
    body: "Libre Franklin"
  },
  targetSize: {
    small: "36px",
    normal: "44px"
  },
  borderRadius: "2px"
}

export default theme
