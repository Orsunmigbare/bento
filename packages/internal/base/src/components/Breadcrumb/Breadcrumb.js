import PropTypes from "prop-types"
import React from "react"
import styled, { css } from "styled-components"
import {
  applyVariants,
  font,
  fontFamily,
  fontWeight,
  spacing,
  color
} from "../../theming"
import { themePropTypes } from "../../constants/themeContants"

const tinycolor = require("tinycolor2")

const Breadcrumb = ({ path }) => (
  <List>
    {path.map(pathItem =>
      pathItem.to ? (
        <Item key={pathItem.name}>
          <ItemButton to={pathItem.to}>{pathItem.name}</ItemButton>
        </Item>
      ) : (
        <ItemCurrent key={pathItem.name}>
          <ItemHeading key={pathItem.name}>{pathItem.name}</ItemHeading>
        </ItemCurrent>
      )
    )}
  </List>
)

Breadcrumb.themeProps = {
  container: {
    name: "BREADCRUMB_CONTAINER",
    description:
      "The container for the crumbs, which also contains custom CSS properties.",
    type: themePropTypes.style
  },
  item: {
    name: "BREADCRUMB_ITEM",
    description: "One single item.",
    type: themePropTypes.style
  },
  itemCurrent: {
    name: "BREADCRUMB_CURRENT_ITEM",
    description: "Active breadcrumb item.",
    type: themePropTypes.style
  },
  itemButton: {
    name: "ITEM_BUTTON",
    description: "The clickable elements in the single item.",
    type: themePropTypes.style
  },
  itemHeading: {
    name: "ITEM_HEADING",
    description: "The non-clickable element in the current item.",
    type: themePropTypes.style
  }
}

const List = styled.ol`
  display: flex;
  flex-wrap: wrap;
  font-size: ${font.base};
  font-family: ${fontFamily.body};
  font-weight: ${fontWeight.normal};

  /* Custom CSS properties – use for customization: */

  --max-width: 150px;
  --bg-color: ${color.white};
  --color: ${color.black};
  --hover-bg-color: ${p =>
    tinycolor(color.primary()(p))
      .darken(0)
      .toString()};
  --hover-color: ${p =>
    tinycolor
      .mostReadable(tinycolor(color.primary()(p)).darken(0), ["#fff"], {
        includeFallbackColors: true,
        level: "AA",
        size: "large"
      })
      .toString()};
  --active-bg-color: ${color.primary};
  --active-color: ${color.white};
  --sep-width: 18px;
  --distance: 3px;
  --line-height: ${font.base};
  --padding-ver: ${spacing.small};
  --padding-hor: ${spacing.small};

  /* Calculations – leave alone: */

  ---height: calc(var(--line-height) + var(--padding-ver) * 2);
  ---push: calc(var(--sep-width) + var(--distance));

  ${applyVariants(Breadcrumb.container)};
`

const itemStyle = css`
  background: var(--bg-color);
  margin-left: var(---push);
  white-space: nowrap;
  position: relative;

  :first-child {
    margin-left: 0;
    > * {
      padding-left: calc(var(--padding-hor) + var(--sep-width) * 0.5);
      ::before {
        content: none;
      }
    }
  }

  :last-child > * {
    padding-right: calc(var(--padding-hor) + var(--sep-width) * 0.5);
    ::after {
      content: none;
    }
  }

  :last-child:not(:only-child)::before {
    border-color: var(--active-bg-color);
    border-left-color: transparent;
  }
`

const Item = styled.li`
  ${itemStyle};
  ${applyVariants(Breadcrumb.item)};
`

const ItemCurrent = styled(Item)`
  ${itemStyle};

  /* Unique for ItemCurrent: */
  ${applyVariants(Breadcrumb.itemCurrent)};
`

const itemInsideStyle = css`
  display: block;
  text-decoration: none;
  max-width: var(--max-width);
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  line-height: var(--line-height);
  color: var(--color);
  padding: var(--padding-ver) var(--padding-hor);
  text-align: center;

  ::before {
    content: "";
    display: block;
    position: absolute;
    width: 0;
    height: 100%;
    top: 0;
    left: calc(var(--sep-width) * -1);
    border: var(--sep-width) solid var(--bg-color);
    border-left-color: transparent;
    border-width: calc(var(---height) / 2) var(--sep-width);
    z-index: -1;
  }

  ::after {
    content: "";
    display: block;
    position: absolute;
    width: 0;
    height: 100%;
    top: 0;
    right: calc(var(--sep-width) * -2);
    border: var(--sep-width) solid transparent;
    border-left-color: var(--bg-color);
    border-width: calc(var(---height) / 2) var(--sep-width);
  }

  :hover,
  :focus {
    outline: none;
    background-color: var(--hover-bg-color);
    color: var(--hover-color);
    text-decoration: none;
    ::before {
      border-color: var(--hover-bg-color);
      border-left-color: transparent;
    }
    ::after {
      border-left-color: var(--hover-bg-color);
    }
  }
`

const ItemButton = styled.a`
  ${itemInsideStyle};
  ${applyVariants(Breadcrumb.themeProps.itemButton)};
`

const ItemHeading = styled.h1`
  ${itemInsideStyle};

  /* Unique for ItemHeading: */

  &,
  :hover {
    background-color: var(--active-bg-color);
    color: var(--active-color);
    ::before {
      border-top-color: var(--active-bg-color);
      border-bottom-color: var(--active-bg-color);
    }
    ::after {
      border-left-color: var(--active-bg-color);
    }
  }

  ${applyVariants(Breadcrumb.themeProps.itemHeading)};
`

Breadcrumb.propTypes = {
  path: PropTypes.array
}

export default Breadcrumb
