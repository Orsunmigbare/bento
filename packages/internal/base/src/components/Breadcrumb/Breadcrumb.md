- Last level will always be _current_ level ('to'-prop can be omitted).
- Key properties are controlled with custom CSS properties assigned to main parent element.

```jsx
<Breadcrumb
  path={[
    {
      name: "First level",
      to: "/"
    },
    {
      name: "Second level",
      to: "/level-2"
    },
    {
      name: "Third level",
      to: "/level-3"
    },
    {
      name: "Fourth level"
    }
  ]}
/>
```
