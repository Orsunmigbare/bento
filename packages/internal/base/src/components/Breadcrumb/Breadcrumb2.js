import PropTypes from "prop-types"
import React from "react"
import styled, { css } from "styled-components"
import { applyVariants, font, fontFamily, spacing, color } from "../../theming"
import { themePropTypes } from "../../constants/themeContants"

const tinycolor = require("tinycolor2")

const Breadcrumb2 = ({ path, ...restProps }) => (
  <BreadcrumbEl {...restProps}>
    {path.map((item, index, arr) =>
      arr.length !== index + 1 ? (
        <Item key={`${item.name}_${index}`}>
          <Link href={item.to}>
            <div>{item.name}</div>
          </Link>
        </Item>
      ) : (
        <Item key={`${item.name}_${index}`}>
          <LastItemContent>
            <div>{item.name}</div>
          </LastItemContent>
        </Item>
      )
    )}
  </BreadcrumbEl>
)

const BreadcrumbEl = styled.ol`
  /* -- parameters: sizes -- */
  --v-padding: ${spacing.tiny};
  --h-padding: ${spacing.small};
  --line-px-height: 24px;
  --edge: 14px;
  --distance: 4px;
  --v-distance: 6px; /* vertical distance â€“ only visible when breadcrumb wraps */

  /* -- parameters: colors -- */
  --c-bg: #ddd;
  --c-txt: ${p =>
    tinycolor
      .mostReadable("#ddd", ["#333"], {
        includeFallbackColors: true,
        level: "AAA",
        size: "small"
      })
      .toString()};
  --c-active-bg: ${color.primary};
  --c-active-txt: ${p =>
    tinycolor
      .mostReadable(color.primary()(p), ["#333", "#fff"], {
        includeFallbackColors: false,
        level: "AA",
        size: "small"
      })
      .toString()};
  --c-hover-bg: ${color.primary};
  --c-hover-txt: ${color.white};

  /* -- calculations -- */
  --height: calc(var(--line-px-height) + var(--v-padding) * 2);

  /* -- styling -- */
  font-family: ${fontFamily.body};
  font-size: ${font.base};
  padding-left: 0;
  padding-right: calc(var(--edge) - var(--distance));
  list-style: none;
  display: flex;
  flex-wrap: wrap;
`

const Item = styled.li`
  background-color: var(--c-bg);
  color: var(--c-txt);
  margin-left: var(--edge);
  margin-right: var(--distance);
  margin-bottom: var(--v-distance);

  &:hover a {
    border-color: var(--c-hover-bg);
  }

  &:not(:last-child):hover {
    background-color: var(--c-hover-bg);
    color: var(--c-hover-txt);
  }

  &:first-child {
    margin-left: 0;
    a::before {
      content: none;
    }
  }

  &:last-child {
    cursor: default;
    margin-right: calc((var(--edge) - var(--distance)) * -1);
    background-color: var(--c-active-bg);
    color: var(--c-active-txt);

    div {
      border-color: var(--c-active-bg);
    }

    ${applyVariants(Breadcrumb2.listItemLast)};
  }

  ${applyVariants(Breadcrumb2.listItem)}

  ${applyVariants(Breadcrumb2.main)};
`

const linkStyle = css`
  display: block;
  text-decoration: none;
  color: inherit;
  border-color: var(--c-bg);
  position: relative;

  ${Item}:not(:first-child) &::before {
    content: "";
    display: block;
    position: absolute;
    height: 100%;
    border: var(--edge) solid;
    border-color: inherit;
    border-left-color: transparent;
    border-top-width: calc(var(--height) / 2);
    border-bottom-width: calc(var(--height) / 2);
    border-right-width: 0;
    box-sizing: border-box;
    left: calc(var(--edge) * -1);
    z-index: -1;
  }

  div {
    display: flex;
    align-items: center;
    line-height: var(--line-px-height);
    padding: var(--v-padding) var(--h-padding);
    white-space: nowrap;

    ${applyVariants(Breadcrumb2.itemText)};
  }

  ${applyVariants(Breadcrumb2.itemAnchor)};
`

const Link = styled.a`
  ${linkStyle};
  ::after {
    content: "";
    display: block;
    position: absolute;
    height: 100%;
    border: var(--edge) solid transparent;
    border-left-color: inherit;
    border-top-width: calc(var(--height) / 2);
    border-bottom-width: calc(var(--height) / 2);
    border-right-width: 0;
    box-sizing: border-box;
    top: 0;
    right: calc(var(--edge) * -1);
  }
`

const LastItemContent = styled.div`
  ${linkStyle};
`

Breadcrumb2.themeProps = {
  main: {
    name: "BREADCRUMB2_CONTAINER",
    description:
      "The main <ol> element, which also contains custom CSS properties.",
    type: themePropTypes.style
  },
  listItem: {
    name: "BREADCRUMB2_ITEM",
    description: "One <li> list item.",
    type: themePropTypes.style
  },
  itemAnchor: {
    name: "BREADCRUMB2_LINK",
    description:
      "Item <a> tag. Will also apply to the last item, which uses a <div>.",
    type: themePropTypes.style
  },
  itemText: {
    name: "BREADCRUMB2_TEXT",
    description: "Item inner <div>.",
    type: themePropTypes.style
  }
}

Breadcrumb2.propTypes = {
  path: PropTypes.array.isRequired,
  separator: PropTypes.bool
}

Breadcrumb2.defaultProps = {
  path: [],
  separator: false
}

export default Breadcrumb2
