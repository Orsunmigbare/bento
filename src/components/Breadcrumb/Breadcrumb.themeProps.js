import { themePropTypes } from "../../theming"

const themeProps = {
  main: {
    name: "BREADCRUMB_CONTAINER",
    description:
      "The main <ol> element, which also contains custom CSS properties.",
    type: themePropTypes.style
  },
  listItem: {
    name: "BREADCRUMB_ITEM",
    description: "One <li> list item.",
    type: themePropTypes.style
  },
  itemAnchor: {
    name: "BREADCRUMB_LINK",
    description:
      "Item <a> tag. Will also apply to the last item, which uses a <div>.",
    type: themePropTypes.style
  },
  itemText: {
    name: "BREADCRUMB_TEXT",
    description: "Item inner <div>.",
    type: themePropTypes.style
  },
  listItemLast: {
    name: "BREADCRUMB_ITEMLAST",
    description: "Item inner <div>.",
    type: themePropTypes.style
  }
}
export default themeProps
