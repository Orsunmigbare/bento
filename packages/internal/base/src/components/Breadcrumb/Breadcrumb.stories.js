import React from "react"
import { storiesOf } from "@storybook/react"
import Breadcrumb from "./Breadcrumb"

storiesOf("Components|Base/Breadcrumb", module)
  .add("No props", () => <Breadcrumb />)
  .add("Strings as input", () => <Breadcrumb path={["Home", "My Car"]} />)
  .add("Object as input", () => (
    <Breadcrumb
      path={[
        {
          name: "First level",
          to: "/"
        },
        {
          name: "Second level",
          to: "/level-2"
        },
        {
          name: "Third level",
          to: "/level-3"
        },
        {
          name: "Fourth level"
        }
      ]}
    />
  ))
  .add("With many objects", () => (
    <Breadcrumb
      path={[
        {
          name: "First level",
          to: "/"
        },
        {
          name: "Second level",
          to: "/"
        },
        {
          name: "Third level",
          to: "/"
        },
        {
          name: "Fourth level",
          to: "/"
        },
        {
          name: "Fifth level",
          to: "/"
        },
        {
          name: "Sixth level",
          to: "/"
        },
        {
          name: "Sevent level",
          to: "/"
        },
        {
          name: "Eight level",
          to: "/"
        }
      ]}
    />
  ))
