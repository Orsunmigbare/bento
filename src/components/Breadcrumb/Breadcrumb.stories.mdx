import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks"
import Breadcrumb from "./Breadcrumb"

<Meta title="Components|Breadcrumb" component={Breadcrumb} />

# Breadcrumb

Breadcrumb is a type of secondary navigation scheme that 
reveals the userâ€™s location in a website or Web application.
Breadcrumbs offer users a way to trace the path back to their original landing point.

```javascript
import {Breadcrumb} from "@staccx/bento"
```

## Standard Breadcrumb

This is a breadcrumb showing the user that they are at the forth page in the line

<Preview>
  <Story name="Standard">
    <Breadcrumb
            path={[
              {
                name: "First page",
                to: "/"
              },
              {
                name: "Second page",
                to: "/level-2"
              },
              {
                name: "Third page",
                to: "/level-3"
              },
              {
                name: "Fourth page"
              }
            ]}
          />
  </Story>
</Preview>

## Multiple Breadcrumbs
Displaying breadcrumbs for mulitple pages. Each of these would be at their own page, showing the uses its
progression through the application

<Preview>
  <Story name="Multiple breadcrumbs">
  <>
    <Breadcrumb
            path={[
              {
                name: "First page",
              }
            ]}
          />
   <Breadcrumb
          path={[
            {
               name: "First page",
               to: "/"
            },
            {
               name: "Second page",
            }
               ]}
            />
   <Breadcrumb
          path={[
            {
               name: "First page",
               to: "/"
            },
            {
               name: "Second page",
               to: "/level-2"
            },
            {
               name: "Third page",
            }
               ]}
            />
       <Breadcrumb
            path={[
            {
             name: "First page",
             to: "/"
            },
            {
              name: "Second page",
              to: "/level-2"
            },
            {
              name: "Third page",
              to: "/level-3"
            },
            {
              name: "Fourth page"
            }
               ]}
            />
          </>
  </Story>
</Preview>

## Props for Breadcrumb

<Props of={Breadcrumb} />
