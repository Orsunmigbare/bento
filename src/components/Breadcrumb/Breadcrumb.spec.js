import React from "react"
import renderer from "react-test-renderer"
import { ThemeProvider } from "styled-components"
import baseTheme from "../../theming/themes/baseTheme"
import Breadcrumb from "./Breadcrumb"

describe("Breadcrumb", () => {
  describe("Snapshots", () => {
    it("no props", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Breadcrumb />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("with path", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Breadcrumb
              path={[
                {
                  name: "First page",
                  to: "/"
                },
                {
                  name: "Second page",
                  to: "/level-2"
                },
                {
                  name: "Third page",
                  to: "/level-3"
                },
                {
                  name: "Fourth page"
                }
              ]}
            />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("empty array", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Breadcrumb path={[]} />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("with child", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Breadcrumb path={[]}>
              <p>test</p>
            </Breadcrumb>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
