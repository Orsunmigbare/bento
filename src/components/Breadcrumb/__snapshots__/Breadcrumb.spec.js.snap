// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb Snapshots empty array 1`] = `
<ol
  className="sc-AykKC hieLbK"
/>
`;

exports[`Breadcrumb Snapshots no props 1`] = `
<ol
  className="sc-AykKC hieLbK"
>
  <li
    className="sc-AykKD duMGmI"
  >
    <div
      className="sc-AykKF fgdujz"
    >
      <div />
    </div>
  </li>
</ol>
`;

exports[`Breadcrumb Snapshots with child 1`] = `
<ol
  className="sc-AykKC hieLbK"
/>
`;

exports[`Breadcrumb Snapshots with path 1`] = `
<ol
  className="sc-AykKC hieLbK"
>
  <li
    className="sc-AykKD duMGmI"
  >
    <a
      className="sc-AykKE bmJenU"
      href="/"
    >
      <div>
        First page
      </div>
    </a>
  </li>
  <li
    className="sc-AykKD duMGmI"
  >
    <a
      className="sc-AykKE bmJenU"
      href="/level-2"
    >
      <div>
        Second page
      </div>
    </a>
  </li>
  <li
    className="sc-AykKD duMGmI"
  >
    <a
      className="sc-AykKE bmJenU"
      href="/level-3"
    >
      <div>
        Third page
      </div>
    </a>
  </li>
  <li
    className="sc-AykKD duMGmI"
  >
    <div
      className="sc-AykKF fgdujz"
    >
      <div>
        Fourth page
      </div>
    </div>
  </li>
</ol>
`;
